### ISSUE TITLE
Refine collection list page

### ISSUE THREAD

**Issue by noheton** [noheton/shepard#251]
***In GitLab by @wiglef on Oct 1, 2024, 13:58 GMT+2:***

# Feature Request

## :goal: Goal

As a `shepard frontend user`I want `a good overview on my collection list`so that `I can navigate to the collection I'm looking for quickly`.

## :medal: Acceptance Criteria

1. Collection Table displays Name, ID, latest Update and Created By 
2. There is an add new collection button that links to https://gitlab.com/dlr-shepard/shepard/-/issues/373
3. Enable the built-in sorting and pagination feature of the data table component
4. There shall be a search functionality that allows to search for name, username, id and collection 
   1. the auto-suggest functionality shall be removed 
   2. if you hit "enter" or "search", the table below is filtered by the search terms 
   3. it would be nice to update the list while typing, select an appropriate mechanism to avoid too many flickering updates 

## :bomb: Out of Scope

1. Collection list has a "Contribution by" Column that displays icons for the contributors
   1. getting icons from AD is desired
   2. fallback if there's no icon
   3. fallback if there is a broken reference
2. There is a favorite button that can be clicked on in - line
   1. the sorting of the list can be selected on the column headers.
   2. the first sorting criterion is always "favorites", the selected sorting criterion is secondary
   3. default secondary sorting criterion is "latest update"

## :bulb: Hints

https://www.figma.com/design/ivBPq71AuUl2ohFnrzHIKF/Visual-Design?node-id=857-8980&t=aSkYhyvnUgwx9ruV-0

## :thinking: Risks & Open Questions

**Assignees:** xit-luk-wer

*Migrated from GitLab: https://gitlab.com/dlr-shepard/shepard/-/work_items/251*

**Comment by noheton** [comment]
***In GitLab by @wiglef on Oct 1, 2024, 13:58 GMT+2:***

added to epic &8

**Comment by noheton** [comment]
***In GitLab by @wiglef on Oct 2, 2024, 10:16 GMT+2:***

added #250 as child task

**Comment by noheton** [comment]
***In GitLab by @wiglef on Oct 2, 2024, 10:17 GMT+2:***

added #249 as child task

**Comment by noheton** [comment]
***In GitLab by @xit-paul-reinhardt on Nov 5, 2024, 09:15 GMT+1:***

I have some open questions:
- How is the favorite mechanism supposed to work? Should it be persisted in shepard or just for one browser/session?
- Responsiveness of the pagination: Is the number of results supposed to change or just the number of pages that can be directly clicked (so e.g. `< 1 ... 3 4 5 ... 15 >` for desktop and `<< < ... 3 4 5 ... > >>` for mobile)
- Is the search functionality already there in the old frontend? If not, can it be moved to a seperate story?

**Comment by noheton** [comment]
***In GitLab by @xit-paul-reinhardt on Jan 16, 2025, 14:35 GMT+1:***

removed child task #250

**Comment by noheton** [comment]
***In GitLab by @xit-paul-reinhardt on Jan 16, 2025, 14:35 GMT+1:***

removed child task #249

**Comment by noheton** [comment]
***In GitLab by @xit-paul-reinhardt on Jan 17, 2025, 09:07 GMT+1:***

mentioned in commit 668dde5daec04f1c04b9d2cf8e42fabc8d722e80

**Comment by noheton** [comment]
***In GitLab by @xit-paul-reinhardt on Jan 17, 2025, 09:07 GMT+1:***

mentioned in commit c554257b560e8d00614eab99bae4977b1a7aa0ed

**Comment by noheton** [comment]
***In GitLab by @xit-paul-reinhardt on Jan 17, 2025, 17:03 GMT+1:***

mentioned in commit 3c7a178f053472a852de1fc3ce707b054b72e5da

**Comment by noheton** [comment]
***In GitLab by @xit-paul-reinhardt on Jan 20, 2025, 09:01 GMT+1:***

I found two issues that increase the effort of this ticket by quite a lot:

1. The pagination of our REST endpoints currently does not return the total number of items. In order to have a proper pagination this is needed to compute the total number of pages.
2. Merging the search term with the result list is not that easy. Until now the normal list uses the `/collections` endpoint (the one with the incomplete pagination mentioned above). This endpoint may take a name parameter, but it will only find exact matches to the name. For the search the `/search` endpoint is used. The search endpoint doesn't take pagination though.

To solve the issues we would need another endpoint in the backend that takes a search term and returns a properly paginated list of matching collections. (Properly paginated meaing a response object with number of total items and current page).

I will mark this as blocked by the container list page since we need similar things there.

**Comment by noheton** [comment]
***In GitLab by @xit-paul-reinhardt on Jan 20, 2025, 09:19 GMT+1:***

marked this issue as blocked by #376

**Comment by noheton** [comment]
***In GitLab by @xit-paul-reinhardt on Jan 20, 2025, 09:46 GMT+1:***

unassigned @xit-paul-reinhardt

**Comment by noheton** [comment]
***In GitLab by @wiglef on Jan 24, 2025, 09:25 GMT+1:***

removed the relation with #376

**Comment by noheton** [comment]
***In GitLab by @xit-mohamed on Feb 5, 2025, 17:03 GMT+1:***

* Button "create new collection" is moving to the middle after click
* icon + font in button "+" should be in textbody (light + dark mode)

**Comment by noheton** [comment]
***In GitLab by @xit-paul-reinhardt on Feb 6, 2025, 09:14 GMT+1:***

I think this will be done anyway after implementing the list according to the design.

**Comment by noheton** [comment]
***In GitLab by @wiglef on Feb 6, 2025, 10:20 GMT+1:***

the search input field that we have on the old collection list now is displaying both chrome- auto fill options as well as bitwarden prompts. 
undefined

that is undesired and i'm pretty sure we have a way to suppress this on our side, at least it does not happen on other input fields that we have elsewhere: 

undefined

**Comment by noheton** [comment]
***In GitLab by @xit-paul-reinhardt on Feb 10, 2025, 13:36 GMT+1:***

added #476 as child task

**Comment by noheton** [comment]
***In GitLab by @xit-paul-reinhardt on Feb 10, 2025, 13:39 GMT+1:***

added #477 as child task

**Comment by noheton** [comment]
***In GitLab by @xit-paul-reinhardt on Feb 14, 2025, 15:21 GMT+1:***

Dev Testing Checklist:

* [ ] ACs:
  * [ ] Collection Table displays Name, ID, latest Update and Created By
  * [ ] There is an add new collection button that links to https://gitlab.com/dlr-shepard/shepard/-/issues/373
  * [ ] Enable the built-in sorting and pagination feature of the data table component
  * [ ] There shall be a search functionality that allows to search for name, username, id and collection
    * [ ] the auto-suggest functionality shall be removed
    * [ ] if you hit "enter" or "search", the table below is filtered by the search terms
    * [ ] it would be nice to update the list while typing, select an appropriate mechanism to avoid too many flickering updates
* [ ] Check implemented endpoints
  * [ ] The /search/collections API works as expected
  * [ ] Errors are properly handled and respond with understandable error codes
  * [ ] No 500 errors are returned

**Comment by noheton** [comment]
***In GitLab by @xit-paul-reinhardt on Feb 14, 2025, 15:46 GMT+1:***

UE testing hints:

* I also adjusted spacing on the container list page, feel free to check that out too
* The search result hint under the search field is not shifting the layout anymore (for both lists). Please check out if it looks nice for you with or without the hint
* The column spacing is now fixed. I allowed all columns to break their content in several lines except for the id. Please check out if you like it or if you want some spacing to be adjusted

**Comment by noheton** [comment]
***In GitLab by @xit-paul-reinhardt on Feb 14, 2025, 16:53 GMT+1:***

mentioned in commit 6bbb40d4c063209c6c58aac009d1fa46ae6043b7

**Comment by noheton** [comment]
***In GitLab by @xit-paul-reinhardt on Feb 14, 2025, 16:53 GMT+1:***

mentioned in commit 8efbcc4a5fa2f08040b9d3701cb4bd8f6a96d95f

**Comment by noheton** [comment]
***In GitLab by @xit-paul-reinhardt on Feb 14, 2025, 17:24 GMT+1:***

mentioned in commit 8cd0ae505b4bad742a8897b576d01c7b94c4abb8

**Comment by noheton** [comment]
***In GitLab by @xit-paul-reinhardt on Feb 14, 2025, 17:24 GMT+1:***

mentioned in commit 173f70260b5ebdd666f0b023cb89e4ca8aedb6d4

**Comment by noheton** [comment]
***In GitLab by @xit-paul-reinhardt on Feb 14, 2025, 17:33 GMT+1:***

mentioned in commit b4316331cfa8ddaf300952302c449fb8142ed095

**Comment by noheton** [comment]
***In GitLab by @xit-paul-reinhardt on Feb 17, 2025, 10:17 GMT+1:***

mentioned in commit ce2f018cb7623c26a3c3f03b0fa6c67cdc2cc12a

**Comment by noheton** [comment]
***In GitLab by @xit-paul-reinhardt on Feb 17, 2025, 10:33 GMT+1:***

mentioned in commit 253eedc06d51478629732007610870d0871e8aac

**Comment by noheton** [comment]
***In GitLab by @xit-paul-reinhardt on Feb 17, 2025, 10:49 GMT+1:***

mentioned in commit 93f035650ba566a9f633af732bf7f9cd8932679f

**Comment by noheton** [comment]
***In GitLab by @wiglef on Feb 17, 2025, 11:27 GMT+1:***

Testing: 
- "Last changed" does not behave as I would expect as a user. I just created a new collection and last changed date stayed empty - would have expected todays date here 
- Rest of ACs look fine for me

**Comment by noheton** [comment]
***In GitLab by @jessicafriedline on Feb 17, 2025, 11:46 GMT+1:***

### ACs:

* [x] Collection Table displays Name, ID, latest Update and Created By

  \--\> @KlaraForner in collections the order is ID, Name, latest Update, Created by. In containers its ID, Name, Created by, Created at. I think that makes sense for the case when we still thought we would get a list of contributors here. I think I would still leave it this way, but wanted to make you aware. 

  \--\> @xit-paul-reinhardt Can the "Last changed" column show the "created at" date, if there were no changes made since? So that it is not empty?

  \--\> Is the last changed column also updated when someone adds a new data object?
* [x] There is an add new collection button that links to https://gitlab.com/dlr-shepard/shepard/-/issues/373

  \--\> the button has a shadow. Can you delete the shadow so the button looks like the others?

  \--\> can you change the icon and font color of the button to "Canvas"? (it will stay white in light mode and the icon won't be black in the dark more)
* [x] Enable the built-in sorting and pagination feature of the data table component
* [x] There shall be a search functionality that allows to search for name, username, id and collection
  * [x] the auto-suggest functionality shall be removed
  * [x] if you hit "enter" or "search", the table below is filtered by the search terms
  * [x] it would be nice to update the list while typing, select an appropriate mechanism to avoid too many flickering updates

    \--\> Nice to have but did and will not make it, right?

### UE Testing hints: 

* I also adjusted spacing on the container list page, feel free to check that out too

  \--\> looks very nice 
* The search result hint under the search field is not shifting the layout anymore (for both lists). Please check out if it looks nice for you with or without the hint

  \--\> thank you, looks great 
* The column spacing is now fixed. I allowed all columns to break their content in several lines except for the id. Please check out if you like it or if you want some spacing to be adjusted

  \--\> looks very good to me! 

### Other

When testing the dark mode I found that if I refresh the page in dark mode it automatically returns to the light mode. Can we change this so that the mode stays the same? Is this a bug ticket?

**Comment by noheton** [comment]
***In GitLab by @xit-luk-wer on Feb 17, 2025, 13:34 GMT+1:***

mentioned in commit c06974f07f4afdd093e49ae011ac6c1f315e9654

**Comment by noheton** [comment]
***In GitLab by @xit-luk-wer on Feb 17, 2025, 13:35 GMT+1:***

- [x] There is an add new collection button that links to https://gitlab.com/dlr-shepard/shepard/-/issues/373

  \--\> the button has a shadow. Can you delete the shadow so the button looks like the others?

  \--\> can you change the icon and font color of the button to "Canvas"? (it will stay white in light mode and the icon won't be black in the dark more)

=\> Fixed in c06974f07f4afdd093e49ae011ac6c1f315e9654

**Comment by noheton** [comment]
***In GitLab by @KlaraForner on Feb 17, 2025, 13:42 GMT+1:***

If it's not a big mess, I would suggest to **swap the columns** "last changed" and "Created by" so that "Last changed" is the last column just as in the Container List. 
If it is more work it's also ok to leave it like that since it's not the complete same column and since maybe some day later the contributor's list might replace the "Created by". 

Apart of that, looking great!

**Comment by noheton** [comment]
***In GitLab by @xit-luk-wer on Feb 17, 2025, 13:45 GMT+1:***

mentioned in commit e8db21e1cd1349c8b41de9d8fdbc5730c6755066

**Comment by noheton** [comment]
***In GitLab by @xit-luk-wer on Feb 17, 2025, 13:47 GMT+1:***

* [x] Collection Table displays Name, ID, latest Update and Created By

  \--\> Can the "Last changed" column show the "created at" date, if there were no changes made since? So that it is not empty?

  ==\> Fixed in e8db21e1cd1349c8b41de9d8fdbc5730c6755066 (This should also fix @wiglef's problem)

  \--\> Is the last changed column also updated when someone adds a new data object?

  ==\> No, it's not

  \--\> Swap **the columns** "last changed" and "Created by" so that "Last changed" is the last column just as in the Container List

  ==\> Fixed in e8db21e1cd1349c8b41de9d8fdbc5730c6755066

**Comment by noheton** [comment]
***In GitLab by @xit-luk-wer on Feb 17, 2025, 13:55 GMT+1:***

mentioned in commit 3241814f04f7495ac7b825352408ef16fa8e0b4f

**Comment by noheton** [comment]
***In GitLab by @xit-luk-wer on Feb 17, 2025, 15:30 GMT+1:***

mentioned in commit f24de5853527bb05871c2b8718c99fa67486a6d3

**Comment by noheton** [comment]
***In GitLab by @xit-luk-wer on Feb 17, 2025, 15:33 GMT+1:***

mentioned in commit 97d305c8747c30021b4f37ac0aae4bb7f9681a94

**Comment by noheton** [comment]
***In GitLab by @xit-luk-wer on Feb 17, 2025, 15:34 GMT+1:***

"When testing the dark mode I found that if I refresh the page in dark mode it automatically returns to the light mode. Can we change this so that the mode stays the same? Is this a bug ticket?"

==\> Fixed in !499

**Comment by noheton** [comment]
***In GitLab by @xit-luk-wer on Feb 17, 2025, 15:46 GMT+1:***

"it would be nice to update the list while typing, select an appropriate mechanism to avoid too many flickering updates"

\--\> Nice to have but did and will not make it, right?

==\> Yes. You either have to press the "Search" button or "Enter" on your keyboard

**Comment by noheton** [comment]
***In GitLab by @xit-luk-wer on Feb 17, 2025, 15:50 GMT+1:***

mentioned in commit 165f2b4235f1d38f9e082189052d367cd37c583c

**Comment by noheton** [comment]
***In GitLab by @xit-luk-wer on Feb 18, 2025, 09:49 GMT+1:***

18.02.2025:

**Problem:**

Displaying the "createdAt" date if a collection has no "updatedAt" date resulted in very confusing behaviour of the sorting. Some collections with a newer date in the "Last changed" column were displayed below other collections with older dates when sorting after the "Last changed" descending. Reason for that was that these collections had no "updatedAt" date and were therefore not considered in the sorting.

**Solution:**

We ( @jessicafriedline , @xit-paul-reinhardt , @xit-luk-wer ) decided to swap the "Last changed" column with a "Created at" column. Now the table is also consistend with the container list page.

Done in !501

**Comment by noheton** [comment]
***In GitLab by @xit-luk-wer on Feb 18, 2025, 09:54 GMT+1:***

mentioned in commit b4ba345c6686a2235f9032c4bfa5956334c12ba5

**Comment by noheton** [comment]
***In GitLab by @jessicafriedline on Feb 18, 2025, 11:30 GMT+1:***

Thank you, I tested the changes just now and everything looks great!

**Comment by noheton** [comment]
***In GitLab by @xit-paul-reinhardt on Feb 19, 2025, 09:05 GMT+1:***

Last thing missing: DEV-Test. See here for the checklist: https://gitlab.com/dlr-shepard/shepard/-/issues/251#note_2348653700

**Comment by noheton** [comment]
***In GitLab by @ZeroOne42 on Feb 19, 2025, 09:34 GMT+1:***

Dev Testing Checklist:

* [ ] ACs:
  * [x] Collection Table displays Name, ID, Created By and Created At (AC changed during implementation, Updated At is replaced by Created At)
  * [x] There is an add new collection button that links to https://gitlab.com/dlr-shepard/shepard/-/issues/373
  * [x] Enable the built-in sorting and pagination feature of the data table component
  * [x] There shall be a search functionality that allows to search for name, username, id and collection (I guess to search for a collection is the same as searching for the name of the collection)
    * [x] the auto-suggest functionality shall be removed (I think that produces errors in the console, but it works)
    * [x] if you hit "enter" or "search", the table below is filtered by the search terms
    * [ ] ~~it would be nice to update the list while typing, select an appropriate mechanism to avoid too many flickering updates~~ (not implemented)
* [ ] Check implemented endpoints
  * [ ] The /search/collections API works as expected
  * [x] Errors are properly handled and respond with understandable error codes
  * [x] No 500 errors are returned

**Comment by noheton** [comment]
***In GitLab by @ZeroOne42 on Feb 19, 2025, 09:41 GMT+1:***

There are a lot of errors in the console: undefined
It is not critical but we should fix it if possible.

**Comment by noheton** [comment]
***In GitLab by @ZeroOne42 on Feb 19, 2025, 16:18 GMT+1:***

When providing a wrong json query the return message is something like that:

{
  "status": 400,
  "exception": "ShepardParserException",
  "message": "could not parse JSON\nUnrecognized token 'Uni': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')\n at [Source: REDACTED (`StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION` disabled); line: 1, column: 4]"
}

We should avoid '\n' in the error message. That will not rendered correctly.

**Comment by noheton** [comment]
***In GitLab by @ZeroOne42 on Feb 19, 2025, 16:28 GMT+1:***

The following request body

```
{
  "searchParams": {
    "query": "{\"property\":\"name\",\"value\":\"Univ\",\"operator\":\"contains\"}"
  }
}
```

returns

```
{
  "searchParams": {
    "query": "{\"property\":\"name\",\"value\":\"Univ\",\"operator\":\"contains\"}"
  },
  "results": [],
  "totalResults": 1
}
```

That is not what I expect. If there is a totalResults of 1, why is the results array empty?

**Comment by noheton** [comment]
***In GitLab by @xit-luk-wer on Feb 20, 2025, 07:41 GMT+1:***

mentioned in commit 62661a16e71cb93fb665cb41a75ab928bf4b62b4

**Comment by noheton** [comment]
***In GitLab by @xit-luk-wer on Feb 20, 2025, 07:42 GMT+1:***

Fixed in 62661a16e71cb93fb665cb41a75ab928bf4b62b4

**Comment by noheton** [comment]
***In GitLab by @xit-luk-wer on Feb 20, 2025, 10:22 GMT+1:***

This issue appeared because the given page in the parameters was 1 and the pagination starts at 0. And since there was only one search result the second page was empty.

For now we will leave the pagination mechanism like this. But this issue is already adressed in the backlog. See --\> https://gitlab.com/dlr-shepard/shepard/-/issues/485

In order to negate this confusion I added a hints to the query params. --\> b750a8bf438c4ecb3e50d19e442562111916424e

**Comment by noheton** [comment]
***In GitLab by @xit-luk-wer on Feb 20, 2025, 10:48 GMT+1:***

mentioned in commit b750a8bf438c4ecb3e50d19e442562111916424e

**Comment by noheton** [comment]
***In GitLab by @xit-luk-wer on Feb 20, 2025, 11:05 GMT+1:***

This issue seems to be caused by the Bitwarden extensions. It also only exists in Chrome and Edge. (Read this https://redirect.github.com/bitwarden/clients/issues/13308)

When using chrome and disableing the bitwarden extensions. I did not get any erros.

**Comment by noheton** [comment]
***In GitLab by @xit-luk-wer on Feb 20, 2025, 14:28 GMT+1:***

mentioned in commit ea6c22c27e8ad78358e344194d063b1498ec9b04

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Feb 24, 2025, 09:54 GMT+1:***

When playing around with the table column sorting options I stumbled across this:

When I first enable coloumn sorting by clicking on the arrow of f.e. the ID column, I see that the query parameters change to `&sortBy={"key":"id","order":"asc"}`, when I click a second time on the same column, the query parameters change as expected to: `&sortBy={"key":"id","order":"desc"}`.
But when I click a third time on it the `sortBy` query parameter is removed and a different ordering is activated.


First Click:

undefined


Second Click:
undefined


Third Click:
undefined

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Feb 24, 2025, 10:12 GMT+1:***

This behavior is intended, therefore closing the comment.
